internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+	if o, ok := err.(*net.OpError); ok {
+		if s, ok := o.Err.(*os.SyscallError); ok {
+			return s.Err == syscall.WSAECONNABORTED
+		}
+	}
+	return false
+}
+
+func (brdg *bridge) recvLoop() {
+	br := bufio.NewReader(brdg.conn)
+	for {
+		id, typ, b, err := readMessage(br)
+		if err != nil {
+			if err == io.EOF || isLocalDisconnectError(err) {
+				brdg.log.Debug("bridge disconnected")
+			} else {
+				brdg.log.WithError(err).Error("bridge read failed")
+			}
+			return
+		}
+		brdg.log.WithFields(logrus.Fields{
+			"payload":    string(b),
+			"type":       typ,
+			"message-id": id}).Debug("bridge receive")
+		switch typ & msgTypeMask {
+		case msgTypeResponse:
+			// Find the request associated with this response.
+			brdg.mu.Lock()
+			call := brdg.rpcs[id]
+			delete(brdg.rpcs, id)
+			brdg.mu.Unlock()
+			if call == nil {
+				brdg.log.WithFields(logrus.Fields{
+					"message-id": id,
+					"type":       typ,
+				}).Error("bridge received unknown rpc response")
+				return
+			}
+			call.complete(json.Unmarshal(b, call.resp))
+
+		case msgTypeNotify:
+			if typ != notifyContainer|msgTypeNotify {
+				brdg.log.WithField("type", typ).Error("bridge received unknown notification message")
+				return
+			}
+			var ntf containerNotification
+			err := json.Unmarshal(b, &ntf)
+			if err != nil {
+				brdg.log.WithError(err).Error("bridge notification unmarshal failed")
+				return
+			}
+			err = brdg.notify(&ntf)
+			if err != nil {
+				brdg.log.WithError(err).Error("bridge notification failed")
+				return
+			}
+		default:
+			brdg.log.WithField("type", typ).Error("bridge received unknown message type")
+			return
+		}
+	}
+}
+
+func (brdg *bridge) sendLoopRoutine() {
+	brdg.sendLoop()
+	// Close the connection so that the reader stops.
+	brdg.conn.Close()
+	// Fail any subsequent RPCs.
+	for call := range brdg.rpcCh {
+		call.complete(errBridgeClosed)
+	}
+}
+
+func (brdg *bridge) writeMessage(buf *bytes.Buffer, enc *json.Encoder, typ msgType, id int64, req interface{}) error {
+	// Prepare the buffer with the message.
+	var h [16]byte
+	binary.LittleEndian.PutUint32(h[:], uint32(typ))
+	binary.LittleEndian.PutUint64(h[8:], uint64(id))
+	buf.Write(h[:])
+	err := enc.Encode(req)
+	if err != nil {
+		return fmt.Errorf("bridge encode: %s", err)
+	}
+	// Update the message header with the size.
+	binary.LittleEndian.PutUint32(buf.Bytes()[4:8], uint32(buf.Len()))
+	// Write the message.
+	brdg.log.WithFields(logrus.Fields{
+		"payload":    string(buf.Bytes()[16:]),
+		"type":       typ,
+		"message-id": id}).Debug("bridge send")
+	_, err = buf.WriteTo(brdg.conn)
+	if err != nil {
+		return fmt.Errorf("bridge write: %s", err)
+	}
+	return nil
+}
+
+func (brdg *bridge) sendLoop() {
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	for call := range brdg.rpcCh {
+		// Prepare the message for the response.
+		brdg.mu.Lock()
+		if brdg.rpcs == nil {
+			brdg.mu.Unlock()
+			call.err = errBridgeClosed
```

Comment:
Can we use `call.complete(errBridgeClosed)` so we have exactly one place that is calling close on the channel. Easier to debug.